opsdroid/connector/twitch/__init__.py

```diff
@@ -0,0 +1,576 @@
+        
+        If this file exists then we just need to read from the file, get the token in the file and
+        attempt to connect to the websockets and subscribe to the Twitch events webhook.
+        
+        """
+        if not os.path.isfile(TWITCH_JSON):
+            _LOGGER.info("No previous authorization data found, requesting new oauth token.")
+            await self.request_oauth_token()
+        else:
+            _LOGGER.info("Found previous authorization data, getting oauth token and attempting to connect.")
+            self.token = self.get_authorization_data()['access_token']
+
+        
+        await self.connect_websocket()
+        
+        self.user_id = await get_user_id(self.default_target, self.token, self.client_id)
+        await self.webhook('follows', 'subscribe')
+        await self.webhook('stream changed', 'subscribe')
+
+    async def reconnect(self):
+        """Attempt to reconnect to the server.
+        
+        Occasionally the Twitch will send a `RECONNECT` command before closing the connection,
+        when this happens we have to connect to the server once more and go through the whole 
+        connection process again.
+        
+        So if the websocket connection drops we will call the `self.connect_websocket` method
+        to attempt to reconnect to the websocket.
+        
+        """
+        _LOGGER.info("Connection to Twitch dropped. Attempting reconnect...")
+        await self.connect_websocket()
+
+    async def listen(self):
+        """Listen method of the connector.
+
+        Every connector has to implement the listen method. When an
+        infinite loop is running, it becomes hard to cancel this task.
+        So we are creating a task and set it on a variable so we can
+        cancel the task.
+
+        """
+        await self.get_messages_loop()
+
+    async def get_messages_loop(self):
+        """Listen for and parse events."""
+        while self.listening:
+            await self._get_messages()
+    
+    async def _get_messages(self):
+        """Receive data from websocket connection.
+        
+        We handle the case where the connection is dropped, if that happens for some
+        reason we attempt to reconnect to the websocket service. This method is being
+        run through the infinite while loop.
+        
+        The message that we get from Twitch contains a lot of metadata, so we are using 
+        regex named groups to get only the data that we need in order to parse a message
+        received.
+        
+        We also need to check if whatever we received from the websocket is indeed a text
+        message that we need to parse, so we use a simple if statement to check if the data
+        received matches a text message.
+        
+        """
+        
+        try:
+            resp = await self.websocket.recv()
+        except websockets.ConnectionClosed:
+            await self.reconnect()
```

Comment:
I would reconnect and then return here. That way you go round the `while self.listening` loop and come back in at the top of the try again. Otherwise if the next `self.websocket.recv` also gets a `websockets.ConnectionClosed` exception it will not be caught.
```suggestion
            await self.reconnect()
            return
        else:
```